src/db_impl.cc

```diff
@@ -1230,14 +1243,29 @@ void TitanDBImpl::OnCompactionCompleted(
         continue;
       }
       if (file->file_state() == BlobFileMeta::FileState::kPendingLSM) {
-        if (delta < 0) {
-          // Cannot happen..
-          ROCKS_LOG_WARN(db_options_.info_log,
-                         "OnCompactionCompleted[%d]: New blob file %" PRIu64
-                         " live size being negative",
-                         compaction_job_info.job_id, file_number);
-          assert(false);
-          delta = 0;
+        if (delta <= 0) {
+          // Because the flushed SST is added to superversion before
+          // `OnFlushCompleted()` is called.
+          // so if there is a concurrent compaction, `OnCompactionCompleted()`
+          // maybe called before `OnFlushCompleted()` is called.
+          // In this case, the state of the blob file generated by the flush is
+          // still `kPendingLSM`, while the blob file size delta
+          // is for the compaction event. So it is possible that delta is
+          // negative.
+          // To make the live data size accurate, here records the delta as a
+          // positive number. And the delta will be subtracted in
+          // `OnFlushCompleted()` later.
+          SubStats(stats_.get(), compaction_job_info.cf_id,
```

Comment:
Maybe we shouldn't add discardable-ratio-level stat when the file is still in `kPendingLSM` state?

Reply:
the discardable-ratio-level is updated when the blob file is added in manifest. But at that time the live_data_size is unknown, so here we need to update the discardable-ratio-level stat.

Reply:
we have talked about it in this PR https://github.com/tikv/titan/pull/157

Reply:
lol, sorry, I feel it weird every time.

Reply:
indeed, will add a todo here.

Reply:
make it not count zero live data size when adding blob file